### DISCUSSION TITLE
Best practice for 1 dockerimage - 4 environments?

### DISCUSSION THREAD

**Discussion by JonShort** [vercel/next.js#22060] [Help]
## TL:DR
Is there a "best practice" approach for pushing one dockerimage through environments which have different - but known at build time - env vars?

## Scenario
I need to push one dockerimage through 4 environments, with different env vars per-environment.

These environments are:
- develop
- integration
- staging
- production

When doing one build, this meant I had to add support for [runtime configuration](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) and disable support for [automatic static optimisation](https://nextjs.org/docs/advanced-features/automatic-static-optimization), which was a compromise, and added a lot of additional configuration.

## Updated approach
At the moment I'm thinking the best approach is to do four next builds when building the dockerimage; then - when spinning up the container - control which build to serve using an environment variable

![Diagram](https://user-images.githubusercontent.com/21317379/107626068-2729c180-6c55-11eb-9d7c-4c04b28eafc9.png)

## Is there a better way?
Is there a better / "best practice" way of progressing one dockerimage through environments?

**Comment by JVMartin** [comment]
As far as we can tell, it's actually not even possible to do this with runtime configuration, as you can see here:
https://github.com/vercel/next.js/discussions/22243

Regarding what you're proposing, other people are instead building a different docker image for each environment:
https://github.com/vercel/next.js/discussions/17641#discussioncomment-110313

Regardless of how you cut it, it is *absolutely unacceptable* to use different docker images or different sub-directories, or ANY other kind of "different compiled assets" in different environments.

We need to have the *same exact artifact* being promoted through test / staging environments into production, otherwise your tests are meaningless.
